timeutil/schedule.go

```diff
@@ -150,34 +154,159 @@ func findNthWeekDay(t time.Time, weekday time.Weekday, nthInMonth uint) time.Tim
 	return t
 }
 
+// findLastWeekDay finds the last occurrence of a given weekday in the month of t
+func findLastWeekDay(t time.Time, weekday time.Weekday) time.Time {
+	n := monthNext(t).Add(-24 * time.Hour)
+	for n.Weekday() != weekday {
+		n = n.Add(-24 * time.Hour)
+	}
+	return n
+}
+
+// matchingWeekdaysInMonth returns the number of occurrences of the weekday of t since
+// the start of the month until t event
+func matchingWeekdaysInMonth(t time.Time) int {
+	month := t.Month()
+	nth := 0
+	for n := t; n.Month() == month; n = n.Add(-7 * 24 * time.Hour) {
+		nth++
+	}
+	return nth
+}
+
 // Match checks if t is within the day span represented by ws.
 func (ws WeekSpan) Match(t time.Time) bool {
 	start, end := ws.Start, ws.End
 	wdStart, wdEnd := start.Weekday, end.Weekday
 
-	if start.Pos != EveryWeek {
-		if start.Pos == LastWeek {
-			// last week of the month
-			if !isLastWeekdayInMonth(t) {
+	weekdayMatch := func(t time.Time) bool {
+		if wdStart <= wdEnd {
+			// single day (mon) or start < end (eg. mon-fri)
+			return t.Weekday() >= wdStart && t.Weekday() <= wdEnd
+		}
+		// wraps around the week end, eg. fri-mon
+		return t.Weekday() >= wdStart || t.Weekday() <= wdEnd
+	}
+
+	if start.Pos == EveryWeek && end.Pos == EveryWeek {
+		// generic weekday match, eg. mon-fri
+		return weekdayMatch(t)
+	}
+
+	// things that use a numbered weekday
+
+	// fun cases, eg (consider the calendar below):
+	//
+	// - mon1-fri, week span, start anchored at 1st Monday 06.08, matches:
+	// 06.08-10.08
+	// - mon-fri2, week span, end anchored at 2nd Friday 10.08, matches:
+	// 06.08-10.08
+	// - fri1-mon, week span, start anchored at 1st Friday 3.08, matches
+	// 03.08-06.08
+	// - mon-fri1, week span, end anchored at 1st Friday 3.08, matches
+	// 30.07-03.08, (crossing the month boundary)
+	// - fri4-thu, week span, end anchored at 4th Friday 27.07, matches
+	// 27.07-02.08, (crossing the month boundary)
+	//
+	//     July 2018            August 2018
+	// Su Mo Tu We Th Fr Sa  Su Mo Tu We Th Fr Sa
+	//  1  2  3  4  5  6  7            1  2  3  4
+	//  8  9 10 11 12 13 14   5  6  7  8  9 10 11
+	// 15 16 17 18 19 20 21  12 13 14 15 16 17 18
+	// 22 23 24 25 26 27 28  19 20 21 22 23 24 25
+	// 29 30 31              26 27 28 29 30 31
+
+	// find out the range of week span, anchor sharing the same month as t
+	startDay, endDay := ws.dateRangeAnchoredAt(t)
+	anchoredAtStart := ws.AnchoredAtStart()
+
+	if t.After(endDay) || t.Before(startDay) {
+		// outside of dates range of the week span, consider edge cases:
```

Comment:
actually this is a good place to have the comment I mentioned above explaining that the edge case here is that there are multiple non-contiguous spans of days in a month that match a range with anchors